torchtune/models/qwen2_5/_prompt_template.py

```diff
@@ -0,0 +1,117 @@
+    """Qwen2.5's chat template.
+
+    Defined in the Jinja template in
+    https://huggingface.co/Qwen/Qwen2.5-7B-Instruct/raw/main/tokenizer_config.json
+
+    Attributes:
+        template: a dictionary mapping role to the
+            prepend and append tags
+
+    Args:
+        add_default_sys_prompt_if_missing (bool): Add Qwen's default system prompt if the messages do not have one.
+            Default: False
+        tools (Optional[List[str]]): Optional JSON schemas for tools that the LLM can access.
+            If provided, the tool schemas will be added to the system prompt according to Qwen's tool instruction format.
+            See https://huggingface.co/blog/unified-tool-use for more information about the JSON tool schema.
+            Default: None
+    """
+
+    template = {
+        "system": ("<|im_start|>system\n", "<|im_end|>\n"),
+        "user": ("<|im_start|>user\n", "<|im_end|>\n"),
+        "assistant": ("<|im_start|>assistant\n", "<|im_end|>\n"),
+    }
+
+    def __init__(
+        self,
+        add_default_sys_prompt_if_missing: bool = False,
+        tools: Optional[List[str]] = None,
+    ):
+        self._add_default_sys_prompt_if_missing = add_default_sys_prompt_if_missing
+        self._tools = tools
+
+    def __call__(
+        self,
+        messages: List[Message],
+        inference: bool = False,
+    ) -> List[Message]:
+        """
+        Format user, assistant, and system messages with appropriate tags.
+
+        Args:
+            messages (List[Message]): a single conversation, structured as a list
+                of `Message` objects
+            inference (bool): Whether the template is being used for inference or not.
+
+        Returns:
+            The formatted list of messages
+        """
+        # Add a system prompt if missing
+        if self._add_default_sys_prompt_if_missing and (
+            not messages or messages[0].role != "system"
+        ):
+            messages.insert(
+                0, Message(role="system", content=DEFAULT_SYS_PROMPT, masked=True)
+            )
+
+        # Add tool instructions to the system prompt
+        if self._tools and messages[0].role == "system":
+            tool_instruction = [TOOL_INSTRUCTION_START]
+            for tool in self._tools:
+                tool_instruction.append(tool)
+            tool_instruction.append(TOOL_INSTRUCTION_END)
+            tool_instruction = "\n".join(tool_instruction)
+            assert messages[0].role == "system"
+            messages[0].content.append({"type": "text", "content": tool_instruction})
+
+        # Add start/end tags to messages
+        formatted_dialogue = []
+        for i, message in enumerate(messages):
+            if message.role == "ipython":
+                if i == 0 or messages[i - 1].role != "ipython":
+                    prepend_tag = f"{self.template['user'][0]}<tool_response>\n"
+                else:
+                    prepend_tag = "\n<tool_response>\n"
```

Comment:
This is interesting.. so we have conditional prepend/append tags depending on contextual information. Feels like this is a gap in our current prompt template abstraction (maybe no way around it though)

Reply:
we already handle similar tags in llama3 in `tokenize_messages`

Reply:
Oh duh good point. Is there a reason we can't move this to the tokenizer then? (Other than the fact that I assume it differs for Qwen2 vs Qwen2.5)

Reply:
See conversation here: https://github.com/pytorch/torchtune/pull/1863#discussion_r1806971190